### PR TITLE
update convolution opencl kernels in dnn module

### PR DESCRIPTION
**Pull request by pengli**
update convolution kernel cache config, cleanup the idlf kernel and extend tuning parameter range. 

```
force_builders=Docs,linux,ocllinux,windows,ocl,macosx,oclmacosx,linuxNoOpt
```

### CODE REVIEW

File: modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp [hunk 1]
```diff
@@ -1826,7 +1824,7 @@ bool OCL4DNNConvSpatial<Dtype>::loadCachedConfig()
         for (size_t i = 0; i < numConfigs; i++)
         {
             std::pair<std::string, std::string> entry(
-                    std::string("Intel(R) Corporation_") + default_kernel_config_intel[2 * i],
+                    ocl::Device::getDefault().vendorName() + "_" + default_kernel_config_intel[2 * i],
```
Comment (alalek) [inline modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: These config numbers for Intel devices only.
`ocl::Device::getDefault()` can point to any OpenCL device, including AMD / NVIDIA.

Did you observe name changing in some Intel OpenCL runtime (platform / version)?

File: modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp [hunk 2]
```diff
@@ -1729,6 +1714,18 @@ void OCL4DNNConvSpatial<float>::setupConvolution(const UMat &bottom,
...
+        double out_z = M_;
+        double k_w = kernel_w_;
+        double k_h = kernel_h_;
+        double k_z = channels_;
+        float elapsedTime = kernelQueue[kernel_index_]->executionTime;
+        double totalFlops = ((k_w*k_h*k_z -1)*2)*(out_w*out_h*out_z)*num_;
+        std::cout << "\tEstimated Gflops:" << (totalFlops * 1e-9)
+                  << std::endl;
+        std::cout << "\tEstimated GFLOPS/S: " << ((totalFlops * 1e-9)*(1000.0/elapsedTime))
+                  << std::endl;
```
Comment (alalek) [inline modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: Please use `CV_LOG_INFO(NULL, ...)` from "opencv2/core/utils/logger.hpp".

Same note for `dbgPrint()` above.

File: modules/dnn/src/ocl4dnn/include/default_kernel_config.hpp [hunk 3]
```diff
@@ -140,77 +140,119 @@ NULL platform behavior
   clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No platform
   clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  No platform
 ********************************************************************************/
-"EU72_k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M112_activ1_eltwise0","7 2 8 2 1 1 8 1 0 ",
```
Comment (alalek) [inline modules/dnn/src/ocl4dnn/include/default_kernel_config.hpp]: Description above declares that OpenCL runtime is "r4.1.61547" - it is outdated / deprecated. Could you collect numbers with fresh Intel OpenCL NEO runtime?

Linux: https://github.com/intel/compute-runtime/releases
Windows: https://downloadcenter.intel.com/product/80939/Graphics-Drivers (or via regular Windows Update)

File: modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp (hunk 2 above)
Comment (pengli) [reply modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: replaced dbgPrint with CV_LOG_INFO

File: modules/dnn/src/ocl4dnn/include/default_kernel_config.hpp (hunk 3 above)
Comment (pengli) [reply modules/dnn/src/ocl4dnn/include/default_kernel_config.hpp]: The cache config is based on latest NEO runtime. The description info is also updated.

File: modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp (hunk 1 above)
Comment (pengli) [reply modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: I just thought "intel" is the default vendor here, but it looks not the case for other users. changed back to old string.

**Comment by pengli** [comment]
@alalek ,  the old convolution cache config is profiled on old opencl runtime and is outdated.  I made the update based on latest NEO. as for performance, I extend auto tuning parameter range, it is helpful for the performance.  I see googlenet  fp32 test improved from 14.5 ms to 13.8 ms on SKL GT2.

**Comment by alalek** [comment]
1.
I observed several kernel verification failures, like:
```
[ERROR:0] test verification failed @ image 0 group 0 out_ch 1 h 0 w 0 got -0 expected 0.885406
[ERROR:0] Kernel IDLF_k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M2048_activ1_eltwise1_2_10_2_1_float_SIMD16 failed verification
[ERROR:0] test verification failed @ image 0 group 0 out_ch 1 h 0 w 0 got -0 expected 0.885406
[ERROR:0] Kernel IDLF_k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M2048_activ1_eltwise1_2_10_1_1_float_SIMD16 failed verification
[ERROR:0] test verification failed @ image 0 group 0 out_ch 1 h 0 w 0 got -0 expected 0.885406
[ERROR:0] Kernel IDLF_k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M2048_activ1_eltwise1_2_10_3_1_float_SIMD16 failed verification
[ INFO:0] Kernel <U_GEMM_LIKE_CONV_k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M2048_activ1_eltwise1_5_2_8_32_float_SIMD8> passed verification
[ INFO:0] Convolution Time:1.06596
```

Also all entries are gone for this configuration (`k1x1_cn512_g1_s1x1_d1x1_b1_in16x16_p0x0_num1_M2048_activ1_eltwise1`) from builtin list (default_kernel_config.hpp).

Could you take a look on this?
- `./bin/opencv_perf_dnn --gtest_filter=DNNTestNetwork.EAST_text_detection/1`
- `OPENCV_OCL4DNN_FORCE_AUTO_TUNING=1` (probably with #11773)
- observed on Linux, i7-6700K, Intel OpenCL NEO 18.23

2.
There is single configuration for both FP32 and FP16 data flow. In general these configurations are not equal (less private memory / cache is used with FP16).
Any thoughts? Should we split them?

File: modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp [hunk 4]
```diff
@@ -1822,13 +1794,25 @@ bool OCL4DNNConvSpatial<Dtype>::loadCachedConfig()
     cv::AutoLock lock(kernelConfigMutex);
     if (!defaultConfigLoaded)
     {
-        const size_t numConfigs = sizeof(default_kernel_config_intel)/sizeof(default_kernel_config_intel[0])/2;
+        std::string prefix = std::string("Intel(R) Corporation_");
+        size_t numConfigs;
+        if (use_half_)
+            numConfigs = sizeof(default_kernel_config_intel_fp16) / sizeof(const char *) / 2;
+        else
```
Comment (alalek) [inline modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: `defaultConfigLoaded` is global variable. So both configs should be initialized at once.

Comment (pengli) [reply modules/dnn/src/ocl4dnn/src/ocl4dnn_conv_spatial.cpp]: @alalek , PR is rebased  on latest 3.4 branch.

**Comment by pengli** [comment]
@alalek , patch is rebased on latest commit, fp32 and fp16 cache config are stored in two arrays and both are loaded at once.  Also the EAST net auto tuning convolution failure is fixed.

**Comment by alalek** [comment]
Results from i7-6700K (Linux 4.15, OpenCL NEO 18.23.10915):
- significant improvement of MobileNet inference (>30%), Inception_v2 (>10%).
- ResNet50 slowdown ~20%
- FP32 slowdown ~10%, FP16 - 20% on AlexNet, GoogLeNet, Inception_5h

BTW, FP16 builtin configurations are less than FP32. Is it correct?

|Name of Test|base|patch|ratio|
|---|:-:|:-:|:-:|
|AlexNet::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|15.062|15.843|0.95|
|AlexNet::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|9.249|11.591|0.80|
|DenseNet_121::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|58.653|58.936|1.00|
|DenseNet_121::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|72.125|72.001|1.00|
|EAST_text_detection::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|98.950|103.408|0.96|
|EAST_text_detection::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|77.425|82.364|0.94|
|ENet::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|22.983|22.216|1.03|
|FastNeuralStyle_eccv16::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|111.126|99.470|1.12|
|GoogLeNet::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|16.735|18.726|0.89|
|GoogLeNet::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|15.959|21.795|0.73|
|Inception_5h::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|18.762|20.523|0.91|
|Inception_5h::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|17.567|23.043|0.76|
|Inception_v2_SSD_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|59.804|50.833|1.18|
|Inception_v2_SSD_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|65.699|57.213|1.15|
|MobileNet_SSD_Caffe::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|21.373|20.951|1.02|
|MobileNet_SSD_Caffe::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|18.079|17.505|1.03|
|MobileNet_SSD_v1_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|23.280|22.143|1.05|
|MobileNet_SSD_v1_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|20.437|18.379|1.11|
|MobileNet_SSD_v2_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|39.390|28.499|1.38|
|MobileNet_SSD_v2_TensorFlow::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|40.567|29.861|1.36|
|OpenFace::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|7.030|7.153|0.98|
|OpenFace::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|8.397|8.772|0.96|
|OpenPose_pose_coco::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|1189.361|1197.590|0.99|
|OpenPose_pose_coco::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|1186.380|1186.233|1.00|
|OpenPose_pose_mpi::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|1171.139|1183.784|0.99|
|OpenPose_pose_mpi::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|1170.678|1170.776|1.00|
|OpenPose_pose_mpi_faster_4_stages::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|864.139|860.567|1.00|
|OpenPose_pose_mpi_faster_4_stages::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|817.631|817.453|1.00|
|ResNet_50::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|37.643|47.296|0.80|
|ResNet_50::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|31.786|41.813|0.76|
|SSD::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|514.519|519.764|0.99|
|SSD::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|346.771|343.540|1.01|
|SqueezeNet_v1_1::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|7.498|7.482|1.00|
|SqueezeNet_v1_1::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|5.829|5.940|0.98|
|YOLOv3::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|372.836|372.442|1.00|
|YOLOv3::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|323.286|323.206|1.00|
|opencv_face_detector::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL)|21.332|20.326|1.05|
|opencv_face_detector::DNNTestNetwork::(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16)|25.150|25.020|1.01|

**Comment by pengli** [comment]
@alalek, i7-6700K is GT2 graphics and has 24EU.  the cache config for 24EU is missing so that you see the performance drop.  I've updated fp32/fp16 cache config for GT2 graphics, so you will see performance improvement on i7-6700K.

**Comment by vpisarev** [comment]
thank you! let's merge it and see what's the performance :+1:

**Comment by alalek** [comment]
@pengli Could you take a look on this issue: #11857 ?

**Pull request merged**